File: osdev_18/src/paging.zig
const std = @import("std");
const isr = @import("isr.zig");
const utils = @import("utils.zig");
const memory = @import("memory.zig");
const Console = @import("driver/Console.zig");
const BitSet = std.bit_set.IntegerBitSet(32);

// https://wiki.osdev.org/Paging#32-bit_Paging_.28Protected_Mode.29
const Page = packed struct {
    present: u1,
    read_write: u1,
    user: u1,
    accessed: u1,
    dirty: u1,
    unused: u7,
    frame: u20,
};

const Table = struct {
    pages: [1024]Page,
};

const Directory = struct {
    tables: [1024]?*Table,
    physical_tables: [1024]u32,
    physical_address: u32,
};

// State for paging
var frames: []BitSet = undefined;
var kernel_directory: *Directory = undefined;

fn firstFrame() ?u32 {
    const amount = frames.len / 32;
    for (frames[0..amount]) |*frame, i| {
        var j: usize = 0;
        while (j < 32) : (j += 1) {
            if (!frame.isSet(j)) {
                frame.set(j);
                return i * 32 + j;
            }
        }
    }
    return null;
}

fn allocateFrame(page: *Page, is_kernel: bool, is_writeable: bool) void {
    if (page.frame == 0) {
        if (firstFrame()) |index| {
            page.present = 1;
            page.read_write = if (is_writeable) 1 else 0;
            page.user = if (is_kernel) 0 else 1;
            page.frame = @truncate(u20, index);
        } else @panic("No free frames!");
    } else return;
}

pub fn getPage(address: u32, create: bool, directory: *Directory) ?*Page {
    const page_address = address / 0x1000;
    const table_index = page_address / 1024;
    if (directory.tables[table_index]) |*table| {
        return &table.*.pages[page_address % 1024];
    } else if (create) {
        var temporary: usize = 0;
        const table = blk: {
            const bytes = memory.malloc(@sizeOf(Table), .page, &temporary);
            break :blk @intToPtr(*Table, bytes);
        };
        directory.tables[table_index] = table;
        directory.physical_tables[table_index] = temporary | 0x7;
        return &table.pages[page_address % 1024];
    } else return null;
}

pub fn switchPageDirectory(directory: *Directory) void {
    asm volatile ("mov %[physical_tables], %%cr3"
        :
        : [physical_tables] "r" (&directory.physical_tables),
    );
    var cr0: u32 = 0;
    asm volatile ("mov %%cr0, %[cr0]"
        : [cr0] "=r" (cr0),
    );
    cr0 |= 0x80000000; // Enable paging!
    asm volatile ("mov %[input], %%cr0"
        :
        : [input] "r" (cr0),
    );
}

pub fn handler(registers: isr.Registers) void {
    const cr2 = asm ("mov %%cr2, %[value]"
        : [value] "=r" (-> u32),
    );
    const present = (registers.error_code & 0x1) == 0;
    const read_write = (registers.error_code & 0x2) > 0;
    const user_mode = (registers.error_code & 0x4) > 0;
    const reserved = (registers.error_code & 0x8) > 0;
    Console.write("Page fault! [ ");
    if (present) Console.write("present ");
    if (read_write) Console.write("read-only ");
    if (user_mode) Console.write("user-mode ");
    if (reserved) Console.write("reserved ");
    Console.write("] at 0x");
    Console.writeHex(cr2);
    Console.write("\n");
    @panic("Page fault");
}

pub fn init() void {
    // Size of physical memory
    const memory_size = 0x1000000;

    // Initialize frames set to 0
    frames = blk: {
        const amount = memory_size / 0x1000;
        const bytes = memory.malloc(@sizeOf(BitSet) * amount, .page, null);
        const items = @intToPtr([*]BitSet, bytes);
        for (items[0..amount]) |*frame|
            frame.* = BitSet.initEmpty();
        break :blk items[0..amount];
    };

    // Create page directory
    kernel_directory = blk: {
        const bytes = memory.malloc(@sizeOf(Directory), .page, null);
        const slice = @intToPtr([*]u8, bytes);
        for (slice[0..@sizeOf(Directory)]) |*byte|
            byte.* = 0;
        break :blk @intToPtr(*Directory, bytes);
    };

    // Identity map physical address to virtual address from 0x0 to end of used memory
    var i: usize = 0;
    while (i < memory.address + 0x1000) : (i += 0x1000) {
        if (getPage(i, true, kernel_directory)) |page|
            allocateFrame(page, false, false);
    }

    // Register page fault handler, then enable paging
    isr.setHandler(14, handler);
    switchPageDirectory(kernel_directory);
}

File: group_36/src/kernel/memory/paging.cpp
#include "paging.h"
#include "memory.h"
#include "../../stdlib/stdlib/cpp/bitset.h"
#include "../../stdlib/stdlib/cpp/exception.h"
#include "../../boot/include/boot.h"
#include "../include/print.h"

void init_paging()
{

    Paging(0x1000000);
}

Paging::Paging(uint32_t mem_end_page)
: frames(mem_end_page, mem_end_page / 0x100)
{
    // printf("\n");
    // print_int(reinterpret_cast<uint32_t>(this));
    // printf("\n");
    // print_int(reinterpret_cast<uint32_t>(this+sizeof(Paging)));
    // printf("\n");

    kernel_directory = (PageDirectory*)kmalloc_a(sizeof(PageDirectory));
    memset(kernel_directory, 0, sizeof(PageDirectory));

    int i = 0;
    while (i < placement_address)
    {
        uint32_t idx = frames.first_available_frame();
        if(kernel_directory->get_page(i, 1)->alloc_frame(idx, 0, 0)){
            frames.set_frame(idx);
        }
        i += 0x1000;
    }
    register_interrupt_handler(ISR14, [](registers_t* regs, void* ctx){
        Paging::page_fault(regs, ctx);
    }, static_cast<void*>(this));

    switch_page_directory(kernel_directory);
    enable_paging();
}

bool Paging::set_directory(PageDirectory *dir) {
    printf("[set-directory] ");
    print_hex(reinterpret_cast<uint32_t>(current_directory));
    printf(" => ");
    print_hex(reinterpret_cast<uint32_t>(dir));
    printf("\n");
    current_directory = dir;
    return false;
}

void Paging::enable_paging() {
    uint32_t cr0; // Define pointer to cr0
    asm volatile("mov %%cr0, %0": "=r"(cr0)); // Move cr0 register to our cr0 pointer
    cr0 |= 0x80000000; // Enable paging!
    asm volatile("mov %0, %%cr0":: "r"(cr0)); // Update cr0 with our cr0 data pointer
}

void Paging::switch_page_directory(PageDirectory *dir) {
    set_directory(dir);
    asm volatile("mov %0, %%cr3":: "r"(&current_directory->tables_physical)); // Move page directory to cr3 register
}

void Paging::page_fault(registers_t* regs, void*)
{
    uint32_t faulting_address;
    asm volatile("mov %%cr2, %0" : "=r" (faulting_address));

    int present   = !(regs->err_code & 0x1); // Page not present
    auto rw = regs->err_code & 0x2;           // Write operation?
    auto us = regs->err_code & 0x4;           // Processor was in user-mode?
    auto reserved = regs->err_code & 0x8;     // Overwritten CPU-reserved bits of page entry?
    auto id = regs->err_code & 0x10;          // Caused by an instruction fetch?

    // Output an error message.
    printf("Page fault: '");
    if (present) {printf("present ");}
    if (rw) {printf("read-only ");}
    if (us) {printf("user-mode ");}
    if (reserved) {printf("reserved ");}
    printf("' at 0x");
    print_hex(faulting_address);
    printf("\n");
}


bool Page::alloc_frame(uint32_t idx , int is_kernel, int is_writeable)
{
    if (frame != 0)
    {
        return false; // Frame was already allocated, return straight away.
    }
    else
    {
        if (idx == (uint32_t)-1)
        {
            // PANIC is just a macro that prints a message to the screen then hits an infinite loop.
            // PANIC("There is no free frames!");
            printf("there are no free frames");
        }

        present = 1; // Mark it as present.
        rw = (is_writeable)?1:0; // Should the page be writeable?
        user = (is_kernel)?0:1; // Should the page be user-mode?
        frame = idx;
    }

    return true;
}

bool Page::free_frame()
{
    uint32_t _frame = 0x0;
    if (!(frame = _frame))
    {
        return false; // The given page didn't actually have an allocated frame!
    }
    else
    {
        frame = 0x0; // Page now doesn't have a frame.

    }
    return true;
}

Page* PageDirectory::get_page(uint32_t address, int make)
{
    // Turn the address into an index.
    address /= 0x1000;
    // Find the page table containing this address.
    uint32_t table_idx = address / NUM_PAGES;
    if (tables[table_idx]) // If this table is already assigned
    {
        return &tables[table_idx]->pages[address % NUM_PAGES];
    }
    else if(make)
    {
        uint32_t tmp;
        tables[table_idx] = (PageTable*)kmalloc_ap(sizeof(PageTable), &tmp);
        memset(tables[table_idx], 0, 0x1000);
        tables_physical[table_idx] = tmp | 0x7; // PRESENT, RW, US.
        return &tables[table_idx]->pages[address % NUM_PAGES];
    }
    else
    {
        return nullptr;
    }
}